File: java/src/main/java/com/example/cote/PGMRS/파괴되지_않은_건물.java
package com.example.cote.PGMRS;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class 파괴되지_않은_건물 {

  private static final int MAX_BOARD_ROWS = 1000;
  private static final int MAX_BOARD_COLS = 1000;
  private static final int MAX_BOARD_DURABILITY = 1000;
  private static final int MAX_SKILLS_LENGTH = 250000;
  private static final int MAX_SKILLS_DEGREE = 100;

  private static int solve(int[][] board, int[][] skills) {
    int row = board.length;
    int col = board[0].length;
    int[][] changes = new int[row + 1][col + 1];

    for (int[] skill : skills) {
      int type = skill[0];
      int r1 = skill[1];
      int c1 = skill[2];
      int r2 = skill[3] + 1;
      int c2 = skill[4] + 1;
      int degree = skill[5];

      if (type == 1) degree = -degree;

      changes[r1][c1] += degree;
      changes[r2][c1] -= degree;
      changes[r1][c2] -= degree;
      changes[r2][c2] += degree;
    }

    for (int y = 0; y < changes.length; y++) {
      for (int x = 0; x < changes[y].length; x++) {
        int left = x > 0 ? changes[y][x - 1] : 0;
        int up = y > 0 ? changes[y - 1][x] : 0;
        int diag = x > 0 && y > 0 ? changes[y - 1][x - 1] : 0;
        changes[y][x] += left + up - diag;
      }
    }

    int safe = 0;

    for (int y = 0; y < board.length; y++) {
      for (int x = 0; x < board[y].length; x++) {
        if (board[y][x] + changes[y][x] > 0) {
          safe++;
        }
      }
    }

    return safe;
  }

  public static void main(String... args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/파괴되지_않은_건물.txt");
    BufferedReader reader = Files.newBufferedReader(p);

    int testCases = Integer.parseInt(reader.readLine());

    while (testCases-- > 0) {
      int[] nums =
          Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

      int[][] board = new int[nums[0]][nums[1]];

      for (int i = 0; i < board.length; i++) {
        board[i] =
            Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
      }

      int n = Integer.parseInt(reader.readLine());
      int[][] skills = new int[n][6];

      for (int i = 0; i < skills.length; i++) {
        skills[i] =
            Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
      }

      int answer = Integer.parseInt(reader.readLine());

      System.out.printf("%d %d\n", answer, solve(board, skills));
    }
  }
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/파괴되지_않은_건물.kt
package com.example.cote.PGMRS;

import java.io.*
import java.nio.file.*

val MAX_BOUND_ROWS = 1000
val MAX_BOUND_COLS = 1000
val MAX_BOARD_DURABILITY = 1000
val MAX_SKILLS_LENGTH = 250000
val MAX_SKILLS_DEGREE = 100

data class Skill (
	val type: Int,
	val r1: Int,
	val c1: Int,
	val r2: Int,
	val c2: Int,
	val degree: Int
	) {
}

fun solve(board: List<List<Int>>, skills: List<List<Int>>): Int {
	val row = board.size;
	val col = board[0].size;

	val changes = MutableList(row + 1) {MutableList(col + 1){0}}

	val skills_ = skills.map {
		Skill(it[0], it[1], it[2], it[3], it[4], it[5])
	}

	for ((type, r1, c1, r2, c2, d) in skills_) {
		val degree = if (type == 1) -d else d

		changes[r1][c1] += degree
		changes[r1][c2 + 1] -= degree
		changes[r2 + 1][c1] -= degree
		changes[r2 + 1][c2 + 1] += degree

	}

	for (i in 0..row) {
		for (j in 0..col) {
			val diag = if (i > 0 && j > 0)  changes[i - 1][j - 1] else 0
			val up = if (i > 0) changes[i - 1][j] else 0
			val left = if (j > 0) changes[i][j - 1] else 0

			changes[i][j] += up + left - diag
		}
	}

	var safe = 0

	for ((i, _) in board.withIndex()) {
		for ((j, durability) in board[i].withIndex()) {
			if (durability + changes[i][j] > 0) {
				safe++
			}
		}
	}

	return safe
}

fun main() {
	val reader = Paths.get(System.getProperty("user.dir") + "/data/파괴되지_않은_건물.txt")
		.run {
			Files.newBufferedReader(this)
		}

	repeat(reader.readLine().toInt()) {
		val nums = reader.readLine().split(" ")
			.map(String::toInt)	
		val board = MutableList(nums[0]){
				reader.readLine().split(" ")
					.map(String::toInt)
					.toMutableList()
			}
		val skills = List(reader.readLine().toInt()) {
			reader.readLine().split(" ")
				.map(String::toInt)
		}
		val answer = reader.readLine().toInt()

		println("$answer ${solve(board, skills)}")
	}
}